src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/MethodResolver.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.StartupHook.MonitorMessageDispatcher.Models;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.Diagnostics.Monitoring.HostingStartup.ParameterCapturing
+{
+    public sealed class MethodResolver
+    {
+        private sealed class DeclaringTypeDescription : IEquatable<DeclaringTypeDescription?>
+        {
+            public string ModuleName { get; set; } = string.Empty;
+            public string ClassName { get; set; } = string.Empty;
+
+            public override bool Equals(object? obj)
+            {
+                return Equals(obj as DeclaringTypeDescription);
+            }
+
+            public bool Equals(DeclaringTypeDescription? other)
+            {
+                return other is not null &&
+                       ModuleName == other.ModuleName &&
+                       ClassName == other.ClassName;
+            }
+
+            public override int GetHashCode()
+            {
+                return HashCode.Combine(ModuleName, ClassName);
+            }
+        }
+
+        private readonly Dictionary<string, List<Module>> _dllNameToModules = new(StringComparer.InvariantCultureIgnoreCase);
+        private readonly Dictionary<DeclaringTypeDescription, List<MethodInfo>> _declaringTypeToMethods = new();
+
+        public MethodResolver()
+        {
+            // Build a lookup table of all viable dll names to their backing reflection Module.
+            IEnumerable<Assembly> assemblies = AppDomain.CurrentDomain.GetAssemblies().Where(
+                assembly => !assembly.ReflectionOnly &&
+                !assembly.IsDynamic);
+
+            foreach (Assembly assembly in assemblies)
+            {
+                foreach (Module module in assembly.GetModules())
+                {
+                    if (!_dllNameToModules.TryGetValue(module.Name, out List<Module>? moduleList))
```

Comment:
This dictionary doesn't represent "dll name to modules". The way it is constructed is "module name to modules". I also think you want a "assembly name to assembly" mapping here and then have the user describe the target methods in terms of (assembly name, type name, method name).

If you want to conflate the term "module" with "assembly", that's... okay, but you have to be internally consistent on the term module actually meaning assembly and thus caching a mapping of assembly names to assemblies, especially since types and methods are uniquely scoped to assemblies, not to modules e.g. you use `Assembly.GetType(name)` to get a `Type`.

FYI, practically, there's a one-to-one mapping between modules and assemblies; the only time you see an assembly to multiple module relationship is if someone gets creative with the compiler (this is not something supported by the project system): https://learn.microsoft.com/en-us/dotnet/framework/app-domains/build-multifile-assembly